### PR TITLE
feat: QUIC datagram transport (RFC 9221/9297)

### PR DESCRIPTION
**Pull request by davidnich**
## Summary
- Add QUIC datagram transport (RFC 9221/9297) with ngtcp2 callbacks for send/receive, per-stream datagram queuing, and transport parameter negotiation
- Add Socket QC API: `submitQuicDatagram()`, `readQuicDatagram()`, `getQuicMaxDatagramSize()`, `isQuicDatagramSupported()`
- Add HttpClientIo datagram methods: `sendDatagram()`, `readDatagram()`, `isDatagramSupported()`, `getMaxDatagramSize()` on `Http3ClientStreamHandle`
- Add `notifyPendingData()` on `HttpClientConnection` to wake the I/O controller after datagram submission
- Fix double-free in `readDatagram()` (BinaryNode ownership of vector data)
- Fix `stream_id=0` validity check (`!exists` vs `!` operator)
- Guard `malloc(0)` edge case for zero-length datagram payloads

## Test plan
- [x] QUIC datagram transport API tests pass (loopback send/receive, negotiation, oversized rejection)
- [x] Existing QUIC/HTTP/3 test suite passes (regression)
- [x] Valgrind clean — 0 errors, 0 definite leaks
- [x] Connect module datagram integration tests pass (19 test cases, 112 assertions) in module-yaml repo

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: lib/QuicSession.cpp [hunk 1]
```diff
@@ -3678,3 +3719,289 @@ void QuicSession::getExtraFds(std::vector<std::pair<int, int>>& extra_fds) const
...
+    size_t varint_len = decodeVarInt(data, datalen, quarter_stream_id);
+    if (varint_len == 0) {
+        printd(2, "recvDatagramCallback(): datagram too short to decode quarter-stream-ID "
+            "(len=%zu)\n", datalen);
+        return 0;  // silently discard malformed datagrams
+    }
+
+    // Convert quarter-stream-ID back to stream ID
+    int64_t stream_id = static_cast<int64_t>(quarter_stream_id * 4);
+
```
[the remark this answers is not in the document]
Comment (davidnich) [reply lib/QuicSession.cpp]: Fixed in df81921f8 — added overflow guard: rejects quarter-stream-ID values that would exceed INT64_MAX/4 before multiplication.

File: lib/QuicSession.cpp [hunk 2]
```diff
@@ -3678,3 +3719,289 @@ void QuicSession::getExtraFds(std::vector<std::pair<int, int>>& extra_fds) const
...
+    // Check remote support
+    const ngtcp2_transport_params* remote_params =
+        ngtcp2_conn_get_remote_transport_params(conn_);
+    if (!remote_params || remote_params->max_datagram_frame_size == 0) {
+        xsink->raiseException("QUIC-DATAGRAM-NOT-SUPPORTED",
+            "remote peer does not support QUIC datagrams (max_datagram_frame_size=0)");
+        return -1;
+    }
+
```
[the remark this answers is not in the document]
Comment (davidnich) [reply lib/QuicSession.cpp:3861]: Fixed in df81921f8 — added validation that stream_id must be a non-negative multiple of 4 per RFC 9297 quarter-stream-ID semantics.

File: include/qore/intern/QuicSession.h [hunk 3]
```diff
@@ -97,6 +97,10 @@ struct QuicStreamInfo {
     //! after being dispatched (so h3RecvDataCallback can route tunnel data)
     bool connect_tunnel_active = false;
 
+    //! RFC 9297: True if this stream is an HTTP/3 datagram anchor (Capsule-Protocol)
+    //! Datagram payloads flow as QUIC DATAGRAM frames using quarter-stream-ID routing.
+    bool datagram_anchor = false;
+
```
[the remark this answers is not in the document]
Comment (davidnich) [reply include/qore/intern/QuicSession.h]: Fixed in df81921f8 — removed the unused `datagram_anchor` field from `QuicStreamInfo`.

File: lib/QuicSession.cpp [hunk 4]
```diff
@@ -3678,3 +3719,289 @@ void QuicSession::getExtraFds(std::vector<std::pair<int, int>>& extra_fds) const
...
+    // Signal waiting readers
+    session->datagram_gen_.fetch_add(1, std::memory_order_release);
+    {
+        std::lock_guard<std::mutex> lg(session->datagram_cv_mtx_);
+    }
+    session->datagram_cv_.notify_all();
+
+    return 0;
```
[the remark this answers is not in the document]
Comment (davidnich) [reply lib/QuicSession.cpp:3833]: Fixed in df81921f8 — wrapped the queue operations and CV signaling in try/catch to prevent C++ exceptions from propagating through ngtcp2's C callback interface.

File: qlib/HttpClientIo/HttpClientStreamHandle.qc [hunk 5]
```diff
@@ -361,6 +361,45 @@ public class HttpClientStreamHandle {
...
+    }
+
+    #! Read the next incoming QUIC datagram on this stream's channel (RFC 9221/9297)
+    /** @param timeout_ms read timeout
+
+        @return the datagram payload, or NOTHING on timeout
+
+        @throw QUIC-DATAGRAM-NOT-SUPPORTED if the protocol does not support QUIC datagrams
+    */
+    *binary readDatagram(timeout timeout_ms = 30s) {
```
[the remark this answers is not in the document]
Comment (davidnich) [reply qlib/HttpClientIo/HttpClientStreamHandle.qc:383]: The `timeout` type in Qore is a dedicated type that accepts duration literals (e.g. `30s`, `500ms`). The `_ms` suffix in the parameter name follows the existing convention in this codebase and is not misleading — it indicates the parameter controls a timeout. No change needed.

File: qlib/HttpClientIo/Http3ClientStreamHandleImpl.qc [hunk 6]
```diff
@@ -63,6 +69,53 @@ public class Http3ClientStreamHandle inherits HttpClientStreamHandle {
...
+    #! Read the next incoming QUIC datagram on this stream's channel (RFC 9221/9297)
+    *binary readDatagram(timeout timeout_ms = 30s) {
+        if (!exists stream_id) {
+            throw "HTTPCLIENT-STATE-ERROR", "stream not yet established";
+        }
+        int sid = h3conn.getSessionId();
+        if (!sid) {
+            throw "HTTPCLIENT-STATE-ERROR", "QUIC session not yet established";
+        }
+        return h3conn.getSocket().readQuicDatagram(sid, stream_id, timeout_ms);
```
[the remark this answers is not in the document]
Comment (davidnich) [reply qlib/HttpClientIo/Http3ClientStreamHandleImpl.qc:96]: Same as above — `timeout_ms` is the established naming convention in this codebase for Qore `timeout` type parameters. No change needed.
